### PR TITLE
net: arp: Add support for gratuitous ARP transmission 

### LINKED ISSUE #57839: net: Implement sending gratuitous ARP once network interface is operational
**Is your enhancement proposal related to a problem? Please describe.**
IoT device discovery by network relies on device able to respond to discovery requests at any give point of time. But when device is in deep sleep (e.g., Wi-Fi6 TWT), the radio is shutoff for longer periods, this causes problems with discovery.

**Describe the solution you'd like**
In order to update the network with IP address of the Device we need to add support in Zephyr networking stack to send out gratuitous ARP once interface is operational and has a valid IP address. This way if a network supports caching discovery details (e.g., proxy ARP) then discovery works even when device is off.

Zephyr already has this feature on RX where ARP table is updated if network sends gratuitous ARP `NET_ARP_GRATUITOUS` for L2 Ethernet.

**Describe alternatives you've considered**
This is just one piece of the solution. Of course this spans both device and network (and possibly cloud/app), so, would need solutions on both sides.

**Additional context**
Matter device discovery with Wi-Fi 6 TWT is our usecase.

### PR DESCRIPTION
**Pull request by rlubos**
Add support for gratuitous ARP transmission by Zephyr network stack.
This allows to prematurely fill the peer ARP table, so there's no need
to send an explicit request when peer needs to send an actual packet.

The gratuitous ARP is send when the network interface is brought up
(joins the network) or a new IP address is added. The gratuitous ARP
request is also sent periodically, with a configurable interval time.
The gratuitous ARP should also be sent whenever MAC address of the
interface is changed, but as Zephyr only allows to do this when
interface is down, this is already covered by the first case (interface
brought up).

Resolves #57839

### CODE REVIEW

File: subsys/net/l2/ethernet/arp.c [hunk 1]
```diff
@@ -466,6 +473,138 @@ static void arp_gratuitous(struct net_if *iface,
...
+
+	notify_all_ipv4_addr(iface);
+}
+
+static void ipv4_event_handler(struct net_mgmt_event_callback *cb,
+			       uint32_t mgmt_event, struct net_if *iface)
+{
+	struct in_addr *ipaddr;
+
+	if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
```
Comment (jukkar) [inline subsys/net/l2/ethernet/arp.c:561]: We need to check also if the interface is VLAN one.
```
@@ -542,7 +542,8 @@ static void iface_event_handler(struct net_mgmt_event_callback *cb,
 {
        ARG_UNUSED(cb);
 
-       if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
+       if (!(net_if_l2(iface) == &NET_L2_GET_NAME(ETHERNET) ||
+             net_eth_is_vlan_interface(iface))) {
                return;
        }
 
@@ -558,7 +559,8 @@ static void ipv4_event_handler(struct net_mgmt_event_callback *cb,
 {
        struct in_addr *ipaddr;
 
-       if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
+       if (!(net_if_l2(iface) == &NET_L2_GET_NAME(ETHERNET) ||
+             net_eth_is_vlan_interface(iface))) {
                return;
        }
 
@@ -583,7 +585,8 @@ static void iface_cb(struct net_if *iface, void *user_data)
 {
        ARG_UNUSED(user_data);
 
-       if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
+       if (!(net_if_l2(iface) == &NET_L2_GET_NAME(ETHERNET) ||
+             net_eth_is_vlan_interface(iface))) {
                return;
        }
```
This requires #71909 which contains the helper function returning this information, I had missed implementing it.

**Review by jukkar — COMMENTED** [review]
Tested it and it works ok, we just need to add some checks for VLAN interfaces.

File: subsys/net/l2/ethernet/Kconfig [hunk 2]
```diff
@@ -73,6 +73,17 @@ config NET_ARP_GRATUITOUS
...
 	  A gratuitous ARP reply is a reply to which no request has been made.
 
+config NET_ARP_GRATUITOUS_TRANSMISSION
+	bool "Transmit gratuitous ARP requests"
+	depends on NET_ARP_GRATUITOUS
+	depends on NET_MGMT_EVENT
+	depends on NET_MGMT_EVENT_INFO
+
+config NET_ARP_GRATUITOUS_INTERVAL
+	int "Time interval [s] between sending gratuitous ARP requests"
```
[the remark this answers is not in the document]
Comment (rlubos) [reply subsys/net/l2/ethernet/Kconfig]: I've updated the description.

File: subsys/net/l2/ethernet/arp.c (hunk 1 above)
Comment (rlubos) [reply subsys/net/l2/ethernet/arp.c]: Done

**Comment by rlubos** [comment]
I've addressed the comments and fixed the build error.

**Review by krish2718 — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**